NitroxClient/Communication/MultiplayerSession/ConnectionState/AwaitingReservationCredentials.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using NitroxClient.Communication.Abstract;
+using NitroxModel.Helper;
+using NitroxModel.MultiplayerSession;
+using NitroxModel.Packets;
+
+namespace NitroxClient.Communication.MultiplayerSession.ConnectionState
+{
+    public class AwaitingReservationCredentials : ConnectionNegotiatingState
+    {
+        public override MultiplayerSessionConnectionStage CurrentStage => MultiplayerSessionConnectionStage.AwaitingReservationCredentials;
+
+        public override void NegotiateReservation(IMultiplayerSessionConnectionContext sessionConnectionContext)
+        {
+            try
+            {
+                ValidateState(sessionConnectionContext);
+
+                string reservationCorrelationId = Guid.NewGuid().ToString();
+                RequestSessionReservation(sessionConnectionContext, reservationCorrelationId);
+                AwaitSessionReservation(sessionConnectionContext, reservationCorrelationId);
+            }
+            catch (Exception)
+            {
+                Disconnect(sessionConnectionContext);
+                throw;
+            }
+        }
+
+        private void RequestSessionReservation(IMultiplayerSessionConnectionContext sessionConnectionContext, string reservationCorrelationId)
+        {
+            IClient client = sessionConnectionContext.Client;
+            PlayerSettings playerSettings = sessionConnectionContext.PlayerSettings;
+            AuthenticationContext authenticationContext = sessionConnectionContext.AuthenticationContext;
+
+            MultiplayerSessionReservationRequest requestPacket = new MultiplayerSessionReservationRequest(reservationCorrelationId, playerSettings, authenticationContext);
+            client.Send(requestPacket);
+        }
+
+        private void AwaitSessionReservation(IMultiplayerSessionConnectionContext sessionConnectionContext, string reservationCorrelationId)
+        {
+            AwaitingSessionReservation nextState = new AwaitingSessionReservation(reservationCorrelationId);
+            sessionConnectionContext.UpdateConnectionState(nextState);
+        }
+
+        private static void ValidateState(IMultiplayerSessionConnectionContext sessionConnectionContext)
+        {
+            ClientIsConnected(sessionConnectionContext);
+            PlayerSettingsIsNotNull(sessionConnectionContext);
+            AuthenticationContextIsNotNull(sessionConnectionContext);
+        }
+
+        private static void ClientIsConnected(IMultiplayerSessionConnectionContext sessionConnectionContext)
+        {
+            if (!sessionConnectionContext.Client.IsConnected)
+            {
+                throw new InvalidOperationException("The client is not connected.");
+            }
+        }
+
+        private static void PlayerSettingsIsNotNull(IMultiplayerSessionConnectionContext sessionConnectionContext)
+        {
+            try
```

Comment:
This is an interesting way of using `Validate.NotNull`...

Reply:
I was torn - but feel that an InvalidOperationException was more appropriate given the context. Figured this was the cleanest way to represent the different concerns.

Reply:
Well, that way you get parameter resolving as well. Looks okay to me, though defining a `NotNull` override throwing an exception of your choosing seems nicer to me; especially since these similar `try-catch` blocks are used in many places.